Fold full-width SIMD-typed indirections

SingleAccretion opened the pull request:
I. e. `OBJ/IND<simd>(ADDR(LCL_VAR<simd>))`.

Some positive (in nature) [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=44369&view=ms.vss-build-web.run-extensions-tab), with a TP win as well.

SingleAccretion committed:
Fold simple SIMD indirs in local morph

SingleAccretion committed:
Fix various downstream bugs

We should really delete the SIMD "InitObj" IR form,
but that change had some unfortunate regressions to
it, so, for now, just make it work reliably.

SingleAccretion committed:
Delete code

SingleAccretion commented:
@dotnet/jit-contrib 

We're getting close to completing the "fold all indirect access in local morph" part of the `ADDR` work.

EgorBo commented:
> positive diffs

doesn't sound good 😄 luckily, they're negative in fact!

EgorBo commented:
Do you want to run some outerloop here? Since over the last few weeks it was constantly in a bad shape we try to run them in all potentially impactful PRs explicitly

SingleAccretion commented:
> Do you want to run some outerloop here? Since over the last few weeks it was constantly in a bad shape we try to run them in all potentially impactful PRs explicitly

Yeah, seems a good idea ([libraries-]stress would be the pipelines I suppose).

> positive diffs

I have a questionable habit of characterizing diffs by their "niceness" 😄.

EgorBo commented:
/azp run runtime-coreclr jitstress, runtime-coreclr libraries-jitstress

SingleAccretion commented:
I looked at the stress failures, all look pre-existing, except one on ARM:
```

Assert failure(PID 61 [0x0000003d], Thread: 119 [0x0077]): Assertion failed 'cls != nullptr && !isInvalidHandle(cls)' in 'Analysis:<MakeAndAssignEnvironments>b__12_0(Scope):this' during 'Morph - Inlining' (IL size 241; hash 0x4aab3218; FullOpts)

    File: /__w/1/s/src/coreclr/jit/_typeinfo.h Line: 235
    Image: /root/helix/work/correlation/dotnet
```
I looked at the dump, and what we have there is a function being inlined, of which the first parameter is a reference type, but when looking at the signature, we get, rather mysteriously, `CORINFO_TYPE_VALUECLASS`. The method table of this type (as returned by `getArgType`) looked uninitialized (assuming I got the right stack location for it, I suppose):
```
0:025> ?? (MethodTable*)0xd0071090
MethodTable * 0xd0071090
   +0x000 m_dwFlags        : 0xf4884050
   +0x004 m_BaseSize       : 0xd0071c90
   +0x008 m_wFlags2        : 0x113c
   +0x00a m_wToken         : 0xd007
   +0x00c m_wNumVirtuals   : 0xdb5c
   +0x00e m_wNumInterfaces : 0xe1d6
   +0x010 debug_m_szClassName : (null) 
   +0x014 m_pParentMethodTable : (null) 
   +0x018 m_pLoaderModule  : 0xffffffff Module
   +0x01c m_pWriteableData : 0x12340000 MethodTableWriteableData
   +0x020 m_pEEClass       : (null) 
   +0x020 m_pCanonMT       : 0
   +0x024 m_pPerInstInfo   : (null) 
   +0x024 m_ElementTypeHnd : 0
   +0x024 m_pMultipurposeSlot1 : 0
   +0x028 m_pInterfaceMap  : 0xdfc265b0 InterfaceInfo_t
   +0x028 m_pMultipurposeSlot2 : 0xdfc265b0
```
Will assume not related as there are no SIMD types on ARM.

Will also hold off on resolving the conflicts until the CI is up.

SingleAccretion commented:
Looks like there are a couple SPMI failures that will need to be investigated.

SingleAccretion committed:
Fix args morphing

Consider a promoted two-field HFA stack argument on ARM64.

This argument needs to be processed by multi-reg morphing,
so that it is transformed into a FIELD_LIST, otherwise it
would end up independently promoted yet appear in the IR
as a LCL_VAR.

That was not happening because the number of stack slots
consumed by such an argument is 1 (8 bytes), and the number
of registers is zero (naturally).

As the fix, use the more obviously correct check based
on "structBaseType" to detect when we need to invoke
multi-reg morphing.

No diffs.

SingleAccretion commented:
> Looks like there are a couple SPMI failures that will need to be investigated.

The issues have been solved and this is once again ready for review.

jakobbotsch commented on src/coreclr/jit/morph.cpp:
```diff
@@ -3255,6 +3244,12 @@ GenTreeCall* Compiler::fgMorphArgs(GenTreeCall* call)
+#if !defined(TARGET_X86)
+                else
+                {
+                    hasMultiregStructArgs = true;
```
Why is this needed here now?

SingleAccretion replied:
It used to be set below:
```c
#if FEATURE_MULTIREG_ARGS
        if (isStructArg)
        {
            if (((arg.AbiInfo.NumRegs + arg.AbiInfo.GetStackSlotsNumber()) > 1) ||
                (isHfaArg && argx->TypeGet() == TYP_STRUCT))
            {
                hasMultiregStructArgs = true;
            }
        }
```
So this is just moving it to here. The benefit it this location is that we don't need to invoke multi-reg morphing if we have already transformed everything into `FIELD_LIST`s.

jakobbotsch commented on src/coreclr/jit/morph.cpp:
```diff
@@ -3443,46 +3438,39 @@ GenTreeCall* Compiler::fgMorphArgs(GenTreeCall* call)
-
             call->gtArgs.SetTemp(&arg, tmp);
-            lvaSetVarAddrExposed(tmp DEBUGARG(AddressExposedReason::TOO_CONSERVATIVE));
+            hasMultiregStructArgs |= ((arg.AbiInfo.ArgType == TYP_STRUCT) && !arg.AbiInfo.PassedByRef);
```
I assume this was a preexisting issue on platforms with by-value struct args?

SingleAccretion replied:
No, as with https://github.com/dotnet/runtime/pull/76745/files#r1009281151, this was just moved.

jakobbotsch reviewed (approved):
LGTM, nice clean up too.
